test/robot/job/device_listener.go

```diff
@@ -0,0 +1,109 @@
+// Copyright (C) 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package job
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/google/gapid/core/app/crash"
+	"github.com/google/gapid/core/log"
+	"github.com/google/gapid/core/os/android/adb"
+	"github.com/google/gapid/core/os/device"
+	"github.com/google/gapid/core/os/device/bind"
+	"github.com/google/gapid/core/os/device/host"
+)
+
+var devReg *bind.Registry
+var device_locks map[string]*sync.Mutex
+var mu sync.Mutex
+
+// BindRegistry creates a new device registry to monitor new Android devices
+// and returnst the context which is bound to the created registry.
+func BindRegistry(ctx context.Context) context.Context {
+	devReg = bind.NewRegistry()
+	ctx = bind.PutRegistry(ctx, devReg)
+	crash.Go(func() {
+		if err := adb.Monitor(ctx, devReg, 15*time.Second); err != nil {
+			log.E(ctx, "adb.Monitor failed: %v", err)
+		}
+	})
+	return ctx
+}
+
+type deviceListener struct {
+	devices  map[string]bind.Device
+	onAdded  func(ctx context.Context, host *device.Instance, target bind.Device)
+	registry *bind.Registry
+}
+
+// OnDevice registers a call back function |onAdded| to be executed when a new
+// Android device is added.
+func OnDevice(ctx context.Context,
+	onAdded func(ctx context.Context, host *device.Instance, target bind.Device)) {
+	dl := &deviceListener{
+		devices:  map[string]bind.Device{},
+		onAdded:  onAdded,
+		registry: bind.GetRegistry(ctx),
+	}
+	dl.registry.Listen(dl)
+}
+
+// OnDeviceAdded implements bind.DeviceListener interface
+func (l *deviceListener) OnDeviceAdded(ctx context.Context, d bind.Device) {
+	mu.Lock()
+	defer mu.Unlock()
+	host := host.Instance(ctx)
+	inst := d.Instance()
+	serial := inst.GetSerial()
+	if _, ok := l.devices[serial]; !ok {
+		log.I(ctx, "Device, added: %s", inst.GetName())
+		l.onAdded(ctx, host, d)
+	} else {
+		l.devices[inst.GetSerial()] = d
+	}
+	if device_locks == nil {
```

Comment:
no underscores in variable names. Please use govet and golint as part of your development environment.

Reply:
Done